lib/srv/desktop/tdp/proto.go

```diff
@@ -1102,6 +1115,252 @@ func decodeSharedDirectoryListResponse(in peekReader) (SharedDirectoryListRespon
+	binary.Write(buf, binary.BigEndian, s)
+	return buf.Bytes(), nil
+}
+
+func decodeSharedDirectoryReadResponse(in peekReader) (SharedDirectoryReadResponse, error) {
+	t, err := in.ReadByte()
+	if err != nil {
+		return SharedDirectoryReadResponse{}, trace.Wrap(err)
+	}
+	if t != byte(TypeSharedDirectoryReadResponse) {
+		return SharedDirectoryReadResponse{}, trace.BadParameter("got message type %v, expected TypeSharedDirectoryReadResponse(%v)", t, TypeSharedDirectoryReadResponse)
+	}
+
+	var completionId, errorCode, readDataLength uint32
+
+	err = binary.Read(in, binary.BigEndian, &completionId)
+	if err != nil {
+		return SharedDirectoryReadResponse{}, trace.Wrap(err)
+	}
+
+	err = binary.Read(in, binary.BigEndian, &errorCode)
+	if err != nil {
+		return SharedDirectoryReadResponse{}, trace.Wrap(err)
+	}
+
+	err = binary.Read(in, binary.BigEndian, &readDataLength)
+	if err != nil {
+		return SharedDirectoryReadResponse{}, trace.Wrap(err)
+	}
+
+	readData := make([]byte, int(readDataLength))
+	if _, err := io.ReadFull(in, readData); err != nil {
+		return SharedDirectoryReadResponse{}, trace.Wrap(err)
+	}
+
+	return SharedDirectoryReadResponse{
+		CompletionID:   completionId,
+		ErrCode:        errorCode,
+		ReadDataLength: readDataLength,
+		ReadData:       readData,
+	}, nil
+}
+
+type SharedDirectoryWriteResponse struct {
+	CompletionID uint32
+	ErrCode      uint32
+	BytesWritten uint32
+}
+
+func (s SharedDirectoryWriteResponse) Encode() ([]byte, error) {
+	fmt.Println("SharedDirectoryWriteResponse: Encode")
+	buf := new(bytes.Buffer)
+	buf.WriteByte(byte(TypeSharedDirectoryWriteResponse))
+	binary.Write(buf, binary.BigEndian, s)
+	return buf.Bytes(), nil
+}
+
+func decodeSharedDirectoryWriteResponse(in peekReader) (SharedDirectoryWriteResponse, error) {
+	t, err := in.ReadByte()
+	if err != nil {
+		return SharedDirectoryWriteResponse{}, trace.Wrap(err)
+	}
+	if t != byte(TypeSharedDirectoryWriteResponse) {
+		return SharedDirectoryWriteResponse{}, trace.BadParameter("got message type %v, expected SharedDirectoryWriteResponse(%v)", t, TypeSharedDirectoryWriteResponse)
+	}
+
+	var res SharedDirectoryWriteResponse
+	err = binary.Read(in, binary.BigEndian, &res)
+	return res, err
+}
+
+type SharedDirectoryWriteRequest struct {
+	CompletionID    uint32
+	DirectoryID     uint32
+	Offset          uint64
+	Path            string
+	PathLength      uint32
+	WriteDataLength uint32
+	WriteData       []byte
+}
+
+func (s SharedDirectoryWriteRequest) Encode() ([]byte, error) {
+	buf := new(bytes.Buffer)
+
+	buf.WriteByte(byte(TypeSharedDirectoryWriteRequest))
+	binary.Write(buf, binary.BigEndian, s.CompletionID)
+	binary.Write(buf, binary.BigEndian, s.DirectoryID)
+	binary.Write(buf, binary.BigEndian, s.Offset)
+	if err := encodeString(buf, s.Path); err != nil {
+		return nil, trace.Wrap(err)
+	}
+	binary.Write(buf, binary.BigEndian, s.WriteDataLength)
+	if _, err := io.WriteString(buf, string(s.WriteData)); err != nil {
```

Comment:
```suggestion
	if _, err := buf.Write(s.WriteData); err != nil {
```

No need to convert to a string when you already have bytes.